components/board/genericlinux/hw_pwm.go

```diff
@@ -0,0 +1,174 @@
+// Package genericlinux is for Linux boards. This particular file is for using sysfs to
+// interact with PWM devices. All of these functions are idempotent: you can double-export a pin or
+// double-close it with no problems.
+package genericlinux
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+type pwmDevice struct {
+	// These values are immutable
+	chipPath string
+	line     int
+
+	mu sync.Mutex
+
+	// These values are mutable
+	periodNs         uint64
+	activeDurationNs uint64
+	isExported       bool
+	isEnabled        bool
+}
+
+func newPwmDevice(chipName string, line int) (*pwmDevice, error) {
+	// There should be a single directory within /sys/devices/platform/<chipName>/pwm/, whose name
+	// is mirrored in /sys/class/pwm. That's the one we want to use.
+	chipDir := fmt.Sprintf("/sys/devices/platform/%s/pwm", chipName)
+	files, err := os.ReadDir(chipDir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, file := range files {
+		if strings.Contains(file.Name(), "pwmchip") && file.IsDir() {
+			chipPath := fmt.Sprintf("/sys/class/pwm/%s", file.Name())
+			return &pwmDevice{chipPath: chipPath, line: line}, nil
+		}
+	}
+	return nil, errors.Errorf("Could not find any PWM device with name %s", chipName)
+}
+
+func writeValue(filepath string, value uint64) error {
+	// The permissions on the file (the third argument) aren't important: if the file needs to be
+	// created, something has gone horribly wrong!
+	return os.WriteFile(filepath, []byte(fmt.Sprintf("%d", value)), 0o600)
+}
+
+func (pwm *pwmDevice) writeChip(filename string, value uint64) error {
+	return writeValue(fmt.Sprintf("%s/%s", pwm.chipPath, filename), value)
+}
+
+func (pwm *pwmDevice) writeLine(filename string, value uint64) error {
+	return writeValue(fmt.Sprintf("%s/pwm%d/%s", pwm.chipPath, pwm.line, filename), value)
+}
+
+// Export tells the OS that this pin is in use, and enables configuration via sysfs.
+func (pwm *pwmDevice) export() error {
+	if pwm.isExported {
+		return nil // Already exported
+	}
+	if err := pwm.writeChip("export", uint64(pwm.line)); err != nil {
+		return err
+	}
+	pwm.isExported = true
+	return nil
+}
+
+// Unexport turns off any PWM signal the pin was providing, and tells the OS that this pin is no
+// longer in use (so it can be reused as an input pin, etc.).
+func (pwm *pwmDevice) unexport() error {
+	if !pwm.isExported {
+		return nil // Already unexported
+	}
+	// If we unexport the pin while it is enabled, it might continue outputting a PWM signal,
+	// causing trouble if you start using the pin for something else.
+	if err := pwm.disable(); err != nil {
+		return err
+	}
+	if err := pwm.writeChip("unexport", uint64(pwm.line)); err != nil {
+		return err
+	}
+	pwm.isExported = false
+	return nil
+}
+
+// Enable tells an exported pin to output the PWM signal it has been configured with.
+func (pwm *pwmDevice) enable() error {
+	if pwm.isEnabled {
```

Comment:
[minor] A debug log would be useful here for knowing when enable has been called redundantly (similarly with disable)

Reply:
I think this would also be helpful for the redundant exports/unexports